crates/iceberg/src/arrow/metadata_column_transformer.rs

```diff
@@ -0,0 +1,282 @@
+        MetadataTransformerBuilder::new(projected_field_ids)
+    }
+
+    /// Applies all registered transformations to the given RecordBatch.
+    ///
+    /// Transformations are applied in the order they were added. Each transformation
+    /// inserts a column at its specified position, so later transformations see the
+    /// updated batch with previously inserted columns.
+    pub(crate) fn apply(&self, mut batch: RecordBatch) -> Result<RecordBatch> {
+        for (transformation, position) in &self.transformations {
+            batch = transformation.apply(batch, *position)?;
+        }
+        Ok(batch)
+    }
+
+    /// Returns true if there are any transformations to apply.
+    pub(crate) fn has_transformations(&self) -> bool {
+        !self.transformations.is_empty()
+    }
+}
+
+impl Default for MetadataTransformer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Builder for constructing a MetadataTransformer from projected field IDs.
+///
+/// This builder analyzes projected field IDs to identify metadata columns (reserved fields)
+/// and builds the appropriate transformations. Reserved fields have special handling and
+/// are inserted into the RecordBatch at their projected positions.
+pub(crate) struct MetadataTransformerBuilder {
+    projected_field_ids: Vec<i32>,
+    file_path: Option<String>,
+}
+
+impl MetadataTransformerBuilder {
+    /// Creates a new MetadataTransformerBuilder for the given projected field IDs.
+    ///
+    /// # Arguments
+    /// * `projected_field_ids` - The list of field IDs being projected, including any reserved fields
+    pub(crate) fn new(projected_field_ids: Vec<i32>) -> Self {
+        Self {
+            projected_field_ids,
+            file_path: None,
+        }
+    }
+
+    /// Sets the file path for the _file metadata column.
+    ///
+    /// # Arguments
+    /// * `file_path` - The file path to use for the _file column
+    ///
+    /// # Returns
+    /// Self for method chaining
+    pub(crate) fn with_file_path(mut self, file_path: String) -> Self {
+        self.file_path = Some(file_path);
+        self
+    }
+
+    /// Builds the MetadataTransformer by analyzing projected field IDs and creating appropriate transformations.
+    ///
+    /// This method:
+    /// 1. Iterates through projected field IDs to find reserved fields
+    /// 2. Calculates the correct position for each reserved field in the final output
+    /// 3. Creates transformations for each reserved field found
+    pub(crate) fn build(self) -> MetadataTransformer {
+        let mut transformations = Vec::new();
+
+        // Iterate through the projected field IDs and check for reserved fields
+        for (position, field_id) in self.projected_field_ids.iter().enumerate() {
+            // Check if this is a reserved field ID for the _file column
+            if *field_id == RESERVED_FIELD_ID_FILE {
+                if let Some(ref path) = self.file_path {
```

Comment:
What happens if there wasn't a file path but there was a `RESERVED_FIELD_ID_FILE`?
Can we handle this dependency differently? E.g. maybe always require file_path? Or something else?